In [1]:
test <- read.csv("../Data/test.csv")
head(test)

A B C D E
1 2 4 4 2 2
2 1 2 2 2 2
3 2 3 4 3 3
4 3 5 5 3 3
5 3 2 4 4 4
6 4 3 3 4 2

In [2]:
tail(test)

A B C D E
397 2 3 5 2 4
398 4 3 3 2 4
399 2 4 4 2 3
400 2 2 5 2 4
401 2 3 4 2 4
402 2 2 4 2 4

In [3]:
str(test)

'data.frame':	402 obs. of  5 variables:
 $ A: int  2 1 2 3 3 4 3 4 4 4 ...
 $ B: int  4 2 3 5 2 3 4 2 4 2 ...
 $ C: int  4 2 4 5 4 3 4 4 5 4 ...
 $ D: int  2 2 3 3 4 4 2 4 3 2 ...
 $ E: int  2 2 3 3 4 2 2 4 4 3 ...


In [4]:
summary(test)

       A               B               C               D              E        
 Min.   :1.000   Min.   :1.000   Min.   :1.000   Min.   :1.00   Min.   :1.000  
 1st Qu.:2.000   1st Qu.:2.000   1st Qu.:3.000   1st Qu.:2.00   1st Qu.:3.000  
 Median :3.000   Median :3.000   Median :4.000   Median :2.00   Median :4.000  
 Mean   :2.734   Mean   :2.908   Mean   :3.622   Mean   :2.51   Mean   :3.386  
 3rd Qu.:3.000   3rd Qu.:4.000   3rd Qu.:4.000   3rd Qu.:3.00   3rd Qu.:4.000  
 Max.   :5.000   Max.   :5.000   Max.   :5.000   Max.   :4.00   Max.   :5.000  

In [5]:
# A 칼럼의 최대값, 최소값, 빈도수 구하기
max(test$A)
min(test$A)
print(table(test$A))

[1] 5

[1] 1


  1   2   3   4   5 
 30 133 156  80   3 


In [6]:
# 각 칼럼의 빈도수 구하기
# test[,1] = test$A
# ncol(test) = 칼럼의 개수
# colnames(test[1]) = 'A'

for(i in 1:ncol(test)){
    cat(colnames(test)[i],"번째 칼럼의 빈도 분석 결과\n")
    print(table(test[,i]))
    cat("-------------------------\n")
}
table(test$A)

A 번째 칼럼의 빈도 분석 결과

  1   2   3   4   5 
 30 133 156  80   3 
-------------------------
B 번째 칼럼의 빈도 분석 결과

  1   2   3   4   5 
 16 150  98 131   7 
-------------------------
C 번째 칼럼의 빈도 분석 결과

  1   2   3   4   5 
  3  74  72 176  77 
-------------------------
D 번째 칼럼의 빈도 분석 결과

  1   2   3   4 
 30 178 153  41 
-------------------------
E 번째 칼럼의 빈도 분석 결과

  1   2   3   4   5 
  8  81 107 160  46 
-------------------------



  1   2   3   4   5 
 30 133 156  80   3 

In [7]:
# 각 칼럼 빈도수의 최대값, 최소값 구하기
for(i in 1:ncol(test)){
    cat(colnames(test)[i],"번째 칼럼 빈도수의 최대값과 최소값\n")
    f<-table(test[,i])
    cat("최대값 :",max(f),"최소값:",min(f),"\n")
    cat("------------------------------\n")
}


A 번째 칼럼 빈도수의 최대값과 최소값
최대값 : 156 최소값: 3 
------------------------------
B 번째 칼럼 빈도수의 최대값과 최소값
최대값 : 150 최소값: 7 
------------------------------
C 번째 칼럼 빈도수의 최대값과 최소값
최대값 : 176 최소값: 3 
------------------------------
D 번째 칼럼 빈도수의 최대값과 최소값
최대값 : 178 최소값: 30 
------------------------------
E 번째 칼럼 빈도수의 최대값과 최소값
최대값 : 160 최소값: 8 
------------------------------


In [10]:
tableMaxMin <- function(x){
    for(i in 1:ncol(x)){
    cat(colnames(x)[i],"번째 칼럼의 빈도 분석 결과")
    print(table(x[,i]))
    cat("\n\n")
    cat(colnames(x)[i],"번째 칼럼 빈도수의 최대값과 최소값\n")
    f<-table(x[,i])
    cat("최대값 :",max(f),"최소값:",min(f),"\n")
    cat("==========================\n\n")
        
    }
    
}

tableMaxMin(test)


A 번째 칼럼의 빈도 분석 결과
  1   2   3   4   5 
 30 133 156  80   3 


A 번째 칼럼 빈도수의 최대값과 최소값
최대값 : 156 최소값: 3 

B 번째 칼럼의 빈도 분석 결과
  1   2   3   4   5 
 16 150  98 131   7 


B 번째 칼럼 빈도수의 최대값과 최소값
최대값 : 150 최소값: 7 

C 번째 칼럼의 빈도 분석 결과
  1   2   3   4   5 
  3  74  72 176  77 


C 번째 칼럼 빈도수의 최대값과 최소값
최대값 : 176 최소값: 3 

D 번째 칼럼의 빈도 분석 결과
  1   2   3   4 
 30 178 153  41 


D 번째 칼럼 빈도수의 최대값과 최소값
최대값 : 178 최소값: 30 

E 번째 칼럼의 빈도 분석 결과
  1   2   3   4   5 
  8  81 107 160  46 


E 번째 칼럼 빈도수의 최대값과 최소값
최대값 : 160 최소값: 8 







---
# 결측치(missing value) 자료 평균 계산


In [11]:
data <- c(10,20,30,40,5,NA,2,NA,3,NA)
data

[1] 10 20 30 40  5 NA  2 NA  3 NA

In [12]:
mean(data)

[1] NA

In [17]:
# NA 갯수 구하기
# true 1 false 0 다 더하면 갯수 나옴
sum(is.na(data))
is.na(data)
# 똑같은 것의 갯수 구하는 것 = 빈도수(table)
table(is.na(data))

[1] 3

[1] FALSE FALSE FALSE FALSE FALSE  TRUE FALSE  TRUE FALSE  TRUE


FALSE  TRUE 
    7     3 

In [19]:
# NA를 제외하고 평균 구하자
# rm = remove : na를 지우기
mean(data, na.rm=T)

# 결측치를 무조건 제거하면 통계량을 얻을 수 없으며, 데이터의 손실로 문제 발생
# -> 데이터프레임에서 한 row가 데이터인데 NA때문에 한 줄이 다 삭제될 수 있음
# NA 개수가 한 칼럼에 30%가 넘으면 그 칼럼을 삭제해야함
# 데이터 수집,저장,처리(다시 수집하러 가는 것),분석

[1] 15.71429

In [21]:
# 방법 1) NA를 0 으로 대체해서 평균값 구하기
mean(ifelse(is.na(data), 0, data))

[1] 11

In [28]:
# 방법 2) NA를 평균으로 대체해서 평균값 구하기
mean(ifelse(is.na(data), mean(data, na.rm=T), data))

[1] 15.71429

In [29]:
# 방법 3) NA를 중앙값으로 대체해서 평균값 구하기
mean(ifelse(is.na(data), median(data, na.rm=T), data))

[1] 14

In [30]:
# 방법 4) NA를 최대값으로 대체해서 평균값 구하기
mean(ifelse(is.na(data), max(data, na.rm=T), data))

[1] 23

In [31]:
# 방법 5) NA를 최소값으로 대체해서 평균값 구하기
mean(ifelse(is.na(data), min(data, na.rm=T), data))

[1] 11.6


> 결측치를 제외한 평균 : 15.7142857142857 \
> 0 으로 대체 : 11\
> 평균으로 대체 : 15.7142857142857\
> 중앙값으로 대체 : 14\
> 최대값으로 대체 : 23\
> 최소값으로 대체 : 11.6\